File: solutions/java/binary-search/step04_bounds/BinarySearch.java
public final class BinarySearch {
    private BinarySearch() {
    }

    public static int binarySearch(int[] array, int target) {
        int left = 0;
        int right = array.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            int value = array[mid];
            if (value == target) {
                return mid;
            }
            if (value < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return -1;
    }
}

File: solutions/python/binary-search/02_found_index/binary_search.py
from typing import Sequence, Any


def binary_search(sequence: Sequence[Any], target: Any) -> int | None:
    left = 0
    right = len(sequence) - 1

    while left <= right:
        mid = (left + right) // 2
        value = sequence[mid]
        if value == target:
            return mid
        if value < target:
            left = mid + 1
        else:
            right = mid - 1

    # "Not found" handling comes next.
    return None

